osu.Game/Screens/Play/HUD/SpectatorList.cs

```diff
@@ -104,22 +107,55 @@ protected override void LoadComplete()
             this.FadeInFromZero(200, Easing.OutQuint);
         }
 
-        private void onSpectatorsChanged(object? sender, NotifyCollectionChangedEventArgs e)
+        private void onWatchingUsersChanged(object? sender, NotifyCollectionChangedEventArgs e)
         {
+            switch (e.Action)
+            {
+                case NotifyCollectionChangedAction.Add:
+                {
+                    for (int i = 0; i < e.NewItems!.Count; i++)
+                        actualSpectators.Add((SpectatorUser)e.NewItems![i]!);
+
+                    break;
+                }
+
+                case NotifyCollectionChangedAction.Remove:
+                {
+                    for (int i = 0; i < e.OldItems!.Count; i++)
+                        actualSpectators.Remove((SpectatorUser)e.OldItems![i]!);
+
+                    break;
+                }
+
+                case NotifyCollectionChangedAction.Reset:
+                {
+                    actualSpectators.Clear();
+                    break;
+                }
+
+                default:
+                    throw new NotSupportedException();
+            }
+
+            removePlayersFromMultiplayerRoom();
+        }
+
+        private void removePlayersFromMultiplayerRoom()
```

Comment:
What happens if the multiplayer client updates first? Not sure entirely how possible this is down this path, but difficult to think about...